repository/git_testing.go

```diff
@@ -42,6 +53,95 @@ func CreateTestRepo(bare bool) *GitRepo {
 	return repo
 }
 
+// setupSigningKey creates a GPG key and sets up the local config so it's used.
+// The key id is set as "user.signingkey". For the key to be found, a `gpg`
+// wrapper which uses only a custom keyring is created and set as "gpg.program".
+// Finally "commit.gpgsign" is set to true so the signing takes place.
+func (repo *GitRepo) setupSigningKey() {
+	config := repo.LocalConfig()
+
+	// Generate a key pair for signing commits.
+	entity, err := openpgp.NewEntity("First Last", "", "fl@example.org", nil)
+	check(err)
+
+	if err := config.StoreString("user.signingkey", entity.PrivateKey.KeyIdString()); err != nil {
+		log.Fatal("failed to set user.signingkey for test repository: ", err)
+	}
+
+	// Armor the private part.
+	privBuilder := &strings.Builder{}
+	w, err := armor.Encode(privBuilder, openpgp.PrivateKeyType, nil)
+	check(err)
+	err = entity.SerializePrivate(w, nil)
+	check(err)
+	err = w.Close()
+	check(err)
+	armoredPriv := privBuilder.String()
+
+	// Armor the public part.
+	pubBuilder := &strings.Builder{}
+	w, err = armor.Encode(pubBuilder, openpgp.PublicKeyType, nil)
+	check(err)
+	err = entity.Serialize(w)
+	check(err)
+	err = w.Close()
+	check(err)
+	armoredPub := pubBuilder.String()
+
+	// Create a custom gpg keyring to be used when creating commits with `git`.
+	keyring, err := ioutil.TempFile("", "keyring")
+	check(err)
+
+	// Import the armored private key to the custom keyring.
+	priv, err := ioutil.TempFile("", "privkey")
+	check(err)
+	_, err = fmt.Fprint(priv, armoredPriv)
+	check(err)
+	err = priv.Close()
+	check(err)
+	err = exec.Command("gpg", "--no-default-keyring", "--keyring", keyring.Name(), "--import", priv.Name()).Run()
+	check(err)
+
+	// Import the armored public key to the custom keyring.
+	pub, err := ioutil.TempFile("", "pubkey")
+	check(err)
+	_, err = fmt.Fprint(pub, armoredPub)
+	check(err)
+	err = pub.Close()
+	check(err)
+	err = exec.Command("gpg", "--no-default-keyring", "--keyring", keyring.Name(), "--import", pub.Name()).Run()
+	check(err)
+
+	// Use a gpg wrapper to use a custom keyring containing GPGKeyID.
+	gpgWrapper := createGPGWrapper(keyring.Name())
+	if err := config.StoreString("gpg.program", gpgWrapper); err != nil {
+		log.Fatal("failed to set gpg.program for test repository: ", err)
+	}
+
+	if err := config.StoreString("commit.gpgsign", "true"); err != nil {
+		log.Fatal("failed to set commit.gpgsign for test repository: ", err)
+	}
+}
+
+// createGPGWrapper creates a shell script running gpg with a specific keyring.
+func createGPGWrapper(keyringPath string) string {
```

Comment:
Can you explain what's the point of this ?

Reply:
When git signs the commit it runs by default `gpg`, or whatever is in the "gpg.program" config entry. We use this `gpg` wrapper so `gpg` can find the temporary keys we create so we can keep them separate from the user's keyring.

Reply:
Is there no way to pass the keys when doing the commit or something ? Deploying a payload like this is not exactly awesome, and doesn't look portable either :-|

Reply:
I think migrating over gogit would help a lot here, but that's some work ...

Reply:
This is not about git, it's about the tool used by git to sign the commit. We need to create a key pair, and we can specify that key just fine by setting `user.signingkey` which is passed as an extra argument to `gpg --status-fd=2 -bsau <user.signingkey>` from what I've noticed.
But the default `gpg` tool will not be able to find it.

Indeed the shell script is not portable, but should be trivial to add the Windows alternative when it's needed.

When migrating to gogit this could be simplified if gogit allows to do it easier?

Reply:
I didn't dig so much in go-git, but it seems to me that it is a relatively complete implementation of git in go, which include support for signatures. It's a bit tricky to find but there is for instance https://github.com/go-git/go-git/blob/master/plumbing/object/commit.go#L348 and https://pkg.go.dev/github.com/go-git/go-git/v5?tab=doc#CommitOptions

It would also have other great benefits: no more calling a external binary, no need to support every version of git anymore, most likely faster, no more constrained by the sometimes awkward git CLI

Reply:
We could start using go-git only for specific operations, such as the commit-tree one, but I don't want to look into this.